tests/app/v2/notifications/test_post_notifications.py

```diff
@@ -1179,3 +1188,443 @@ def test_post_notification_returns_400_when_get_json_throws_exception(client, sa
+    assert error_json["errors"] == [
+        {
+            "error": "BadRequestError",
+            "message": "You cannot send to these recipients because your service is in trial mode. You can only send to members of your team and your safelist.",
+        }
+    ]
+
+
+def test_post_bulk_flags_not_enough_remaining_messages(client, notify_db, notify_db_session, mocker):
+    service = create_service(message_limit=10)
+    template = sample_template(notify_db, notify_db_session, service=service, template_type="email")
+    messages_count_mock = mocker.patch("app.v2.notifications.post_notifications.fetch_todays_total_message_count", return_value=9)
+    data = {
+        "name": "job_name",
+        "template_id": template.id,
+        "csv": rows_to_csv([["email address"], ["foo@example.com"], ["bar@example.com"]]),
+    }
+
+    response = client.post(
+        "/v2/notifications/bulk",
+        data=json.dumps(data),
+        headers=[("Content-Type", "application/json"), create_authorization_header(service_id=template.service_id)],
+    )
+
+    assert response.status_code == 400
+    error_json = json.loads(response.get_data(as_text=True))
+    assert error_json["errors"] == [
+        {
+            "error": "BadRequestError",
+            "message": "You only have 1 remaining messages before you reach your daily limit. You've tried to send 2 messages.",
+        }
+    ]
+    messages_count_mock.assert_called_once()
+
+
+@pytest.mark.parametrize("data_type", ["rows", "csv"])
+def test_post_bulk_flags_rows_with_errors(client, notify_db, notify_db_session, data_type):
+    template = sample_template(notify_db, notify_db_session, template_type="email", content="Hello ((name))")
+    data = {"name": "job_name", "template_id": template.id}
+    rows = [
+        ["email address", "name"],
+        ["foo@example.com", "Foo"],
+        ["bar@example.com"],
+        ["nope", "nope"],
+        ["baz@example.com", ""],
+        ["baz@example.com", " "],
+    ]
+    if data_type == "csv":
+        data["csv"] = rows_to_csv(rows)
+    else:
+        data["rows"] = rows
+
+    response = client.post(
+        "/v2/notifications/bulk",
+        data=json.dumps(data),
+        headers=[("Content-Type", "application/json"), create_authorization_header(service_id=template.service_id)],
+    )
+
+    assert response.status_code == 400
+    error_json = json.loads(response.get_data(as_text=True))
+    assert error_json["errors"] == [
+        {
+            "error": "BadRequestError",
+            "message": "Some rows have errors. Row 1 - `name`: Missing. Row 2 - `email address`: invalid recipient. Row 3 - `name`: Missing. Row 4 - `name`: Missing.",
+        }
+    ]
+
+
+@pytest.mark.parametrize("data_type", ["rows", "csv"])
+@pytest.mark.parametrize("is_scheduled", [True, False])
+@pytest.mark.parametrize("use_sender_id", [True, False])
+def test_post_bulk_creates_job_and_dispatches_celery_task(
```

Comment:
This is the success case. Testing with rows/csv, scheduled or not, custom sender or not.